### PR TITLE
Refactor search: sliding window as library 

### PR DESCRIPTION
**Pull request by eaasna**
Resolves https://github.com/eaasna/sliding-window/issues/2

And resolves https://github.com/eaasna/sliding-window/issues/20 

- [x] divide search into smaller manageable functions
- [x] extract the main `worker` function that queries the IBF for a batch of sequences so that it can be used by external apps
- [x] separate `sliding_window::app` namespace

### CODE REVIEW

File: include/sliding_window/search/do_parallel.hpp [hunk 1]
```diff
@@ -4,14 +4,16 @@
...
+
 namespace sliding_window
 {
 
 template <typename t>
 inline void do_parallel(t && worker, size_t const num_records, size_t const threads, double & compute_time)
 {
     auto start = std::chrono::high_resolution_clock::now();
-    std::vector<decltype(std::async(std::launch::async, worker, size_t{}, size_t{}))> tasks;
+    std::vector<decltype(std::async(std::launch::async, worker, size_t{}, size_t{}))> tasks; // std::async returns a std::future
```
Comment (eaasna) [inline include/sliding_window/search/do_parallel.hpp:16]: Hi @marehr please ignore the rest of the PR for now (it's messy). 

But my question would be about this parallelization that is already there for my application. I take a set of query reads assign them to threads and execute in parallel. I execute the `worker` method which finds corresponding bins to each read. Is this also something that should be taken care of the `slotted_cart_queue`? 

Or would we want the `slotted_cart_queue` to only be used for the next step i.e bulk execution of the STELLAR read mapping?

File: include/sliding_window/search/search_setup.hpp [hunk 2]
```diff
@@ -0,0 +1,262 @@
...
+            seqan3::sequence_file_input<dna4_traits, seqan3::fields<seqan3::field::id, seqan3::field::seq>> fin{arguments.query_file};
+            using record_type = typename decltype(fin)::record_type;
+            std::vector<record_type> records{};
+
+            auto const threshold_data = make_threshold_data(arguments);
+        }
+};
+
+// auto worker = [&](size_t const start, size_t const end)
+auto worker(size_t const start, size_t const end, search_arguments const & arguments, search_setup const & setup)
```
Comment (eaasna) [inline include/sliding_window/search/search_setup.hpp:166]: Here template parameter deduction for `search_setup` does not work

File: include/sliding_window/search/search_setup.hpp [hunk 3]
```diff
@@ -0,0 +1,262 @@
...
+        search_setup(query_result const &) = default;
+        search_setup & operator=(search_setup const &) = default;
+        search_setup(query_result &&) = default;
+        search_setup & operator=(search_setup &&) = default;
+        ~search_setup() = default;
+*/
+        template <bool compressed>
+        search_setup(search_arguments const &arguments, search_time_statistics & time_statistics)
+        {
+            constexpr seqan3::data_layout ibf_data_layout = compressed ? seqan3::data_layout::compressed : seqan3::data_layout::uncompressed;
```
Comment (eaasna) [inline include/sliding_window/search/search_setup.hpp:147]: This only works if `compressed` is a template parameter

File: include/sliding_window/search/search_setup.hpp [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+
+//-----------------------------
+//
+// Search a batch of reads (start; end) in the IBF
+//
+// TODO: pass slice of records directly instead of start, end and records
+//-----------------------------
+template <typename ibf_t, typename rec_vec_t>
+auto worker(size_t const start, size_t const end, ibf_t const & ibf, search_arguments const & arguments, 
+                                                                    rec_vec_t const & records, threshold const & threshold_data)
```
Comment (marehr) [inline include/sliding_window/search/search_setup.hpp:135]: whitespace

File: include/sliding_window/search/search_setup.hpp [hunk 5]
```diff
@@ -0,0 +1,231 @@
...
+        // window_span_begin 	=     0		4	   5
+        // window_span_end 	=     10	11	   12
+        //
+        // minimiser	span
+        // AAAA		CGCAAAACGCG
+        // AAAC		AAACGCGG
+        // AACG		AACGCGGC
+        //
+        //-----------------------------
+        for (auto begin : begin_vector)
```
Comment (marehr) [inline include/sliding_window/search/search_setup.hpp:217]:
```suggestion
        for (auto && begin : begin_vector)
```

File: src/CMakeLists.txt [hunk 6]
```diff
@@ -51,3 +51,6 @@ target_link_libraries ("${PROJECT_NAME}_lib" INTERFACE "${PROJECT_NAME}_search_l
 # Sliding window executable
 add_executable ("${PROJECT_NAME}" sliding_window_main.cpp)
 target_link_libraries ("${PROJECT_NAME}" PUBLIC "${PROJECT_NAME}_lib")
+
+# Enable debugging in VisualStudio
+set_target_properties ("${PROJECT_NAME}" PROPERTIES COMPILE_FLAGS "-g")
```
Comment (marehr) [inline src/CMakeLists.txt:56]: Maybe remove this

File: src/sliding_window_search.cpp [hunk 7]
```diff
@@ -1,26 +1,65 @@
...
 {
 
-void sliding_window_search(search_arguments const & arguments)
+//-----------------------------
+//
+// Setup IBF and launch multithreaded search.
+//
+//-----------------------------
+template <bool compressed>
+auto run_program(search_arguments const &arguments, search_time_statistics & time_statistics)
```
Comment (marehr) [inline src/sliding_window_search.cpp:12]: is this void?

File: src/sliding_window_search.cpp [hunk 8]
```diff
@@ -1,26 +1,65 @@
...
-    else
+        load_ibf(ibf, arguments, time_statistics.ibf_io_time);
+    };
+
+    auto cereal_handle = std::async(std::launch::async, cereal_worker);
+
+    seqan3::sequence_file_input<dna4_traits, seqan3::fields<seqan3::field::id, seqan3::field::seq>> fin{arguments.query_file};
+    using record_type = typename decltype(fin)::record_type;
+    std::vector<record_type> records{};
+    using rec_vec_t = decltype(records);
```
Comment (marehr) [inline src/sliding_window_search.cpp:28]: You basically know the type, I guess.

**Review by marehr — APPROVED** [review]

File: include/sliding_window/search/query_result.hpp [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+    query_result(query_result const &) = default;
+    query_result & operator=(query_result const &) = default;
+    query_result(query_result &&) = default;
+    query_result & operator=(query_result &&) = default;
+    ~query_result() = default;
+
+    query_result(std::string const query_id, std::set<size_t> query_hits)
+    {
+        id = query_id;
+        bin_hits = query_hits;
```
Comment (eaasna) [inline include/sliding_window/search/query_result.hpp:24]: use move for query_hits set, could it be unordered set instead?

File: include/sliding_window/search/query_result.hpp [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+    query_result & operator=(query_result &&) = default;
+    ~query_result() = default;
+
+    query_result(std::string const query_id, std::set<size_t> query_hits)
+    {
+        id = query_id;
+        bin_hits = query_hits; 
+    }
+
+    std::set<size_t> get_hits()
```
Comment (eaasna) [inline include/sliding_window/search/query_result.hpp:27]: avoid copy!!!

**Pull request merged**